### PR TITLE
Chore: (Docs) Adds Storybook builder documentation

### PR DESCRIPTION
**Pull request by jonniebigodes**
With this pull request the documentation is updated to feature Storybook's Builder API and Vite Builder.

What was done:
- Added documentation on Vite Builder, Builder API
- Added snippets for the above
- Updated TOC file to reflect changes
- Updated Webpack documentation.

Feel free to provide feedback. @IanVS If you'd be so kind as to check this pull request and provide some insights on the Vite builder front I'd appreciate it. Thanks in advance

### CODE REVIEW

File: docs/builders/webpack.md [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+```shell
+
+## Development mode
+yarn start-storybook --debug-webpack
+
+## Production mode
+yarn build-storybook --debug-webpack
+```
+
+### Bundle splitting
```
Comment (winkerVSbecks) [inline docs/builders/webpack.md:54]: Should we just rename this to code splitting? That's the more common name and we refer to it as code splitting everywhere else.

File: docs/builders/overview.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
+---
+title: 'Builders'
+---
+
+Storybook, at its core, is a build tool. It provides pluggable builders to help match the application's development environment, making it easy to compile your code—Javascript, CSS, and MDX into an executable bundle, updating in the browser in real-time.
```
Comment (winkerVSbecks) [inline docs/builders/overview.md:5]:
```suggestion
Storybook, at its core, is powered by builders such as Webpack and Vite. These builders spin up a development environment,  compile your code—Javascript, CSS, and MDX—into an executable bundle and update the browser in real-time.
```

**Review by winkerVSbecks — COMMENTED** [review]
@jonniebigodes really nice work. I left a few minor suggestions.

**Review by shilman — COMMENTED** [review]
Does this PR need a corresponding PR on the frontpage repo to redirect `configure/webpack` to `builders/webpack`?

**Comment by winkerVSbecks** [comment]
@shilman yes, good catch

File: docs/builders/builder-api.md [hunk 3]
```diff
@@ -0,0 +1,163 @@
+---
+title: 'Builder API'
+---
+
+Storybook is architected to support multiple builders, including [Webpack](https://webpack.js.org/), [Vite](https://vitejs.dev/), and [ESBuild](https://esbuild.github.io/). The builder API is the set of interfaces to add a new builder to Storybook.
```
Comment (domyen) [inline docs/builders/builder-api.md:5]:
```suggestion
Storybook is architected to support multiple builders, including [Webpack](https://webpack.js.org/), [Vite](https://vitejs.dev/), and [ESBuild](https://esbuild.github.io/). The builder API is the set of interfaces you can use to add a new builder to Storybook.
```

File: docs/builders/builder-api.md [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+<CodeSnippets
+  paths={[
+    'common/storybook-builder-api-interface.ts.mdx',
+  ]}
+/>
+
+<!-- prettier-ignore-end -->
+
+In development mode, the `start` API call is responsible for initializing the development server to monitor the file system for changes (for example, components, stories) and execute a hot module reload in the browser.
```
Comment (domyen) [inline docs/builders/builder-api.md:50]:
```suggestion
In development mode, the `start` API call is responsible for initializing the development server to monitor the file system for changes (for example, components and stories) then execute a hot module reload in the browser.
```

File: docs/builders/builder-api.md [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+It also provides a **bail** function to allow the running process to end gracefully, either via user input or error.
+
+In production, the `build` API call is responsible for generating a static Storybook build, storing it by default in the `storybook-static` directory if no additional configuration is provided. The generated output should contain everything the user needs to view its Storybook by opening either the `index.html` or `iframe.html` in a browser with no other processes running.
```
Comment (domyen) [inline docs/builders/builder-api.md:53]: Should we add the whole path here to indicate that it is a directory?

For instance: `.storybook/storybook-static` or `/storybook-static` 
(Not sure where `storybook-static` gets put by default.)

File: docs/builders/builder-api.md [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+<!-- prettier-ignore-end -->
+
+### Generate source code snippets
+
+Storybook annotates components and stories with additional metadata related to their inputs to automatically generate interactive controls and documentation. Currently, this is provided via Webpack loaders/plugins. The builder must re-implement this to support those features.
+
+### Generate a static build
+
+One of Storybook's core features it's the ability to generate a static build that can be [published](../sharing/publish-storybook.md) into any hosting service. The builder must also be able to provide a similar mechanism. For example:
```
Comment (domyen) [inline docs/builders/builder-api.md:111]:
```suggestion
One of Storybook's core features it's the ability to generate a static build that can be [published](../sharing/publish-storybook.md) to a web hosting service. The builder must also be able to provide a similar mechanism. For example:
```

File: docs/builders/builder-api.md [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+<!-- prettier-ignore-end -->
+
+### HMR support
+
+While running in development mode, the builder's development server must be able to reload the page once a change happens, either in a story, component, or helper function.
+
+### More information
+
+This area is under rapid development, so it's barely documented and subject to change. If you are interested in writing your builder, we encourage you to check [webpack](https://github.com/storybookjs/storybook/tree/next/lib/builder-webpack4), [Vite](https://github.com/storybookjs/builder-vite), and Modern Web's [dev-server-storybook](https://github.com/modernweb-dev/web/blob/master/packages/dev-server-storybook/src/serve/storybookPlugin.ts) source code. Additionally, we have a wonderful contributor community on the [Storybook Discord](https://discord.gg/storybook) if you have questions. Ping us in the #contributing channel.
```
Comment (domyen) [inline docs/builders/builder-api.md:157]:
```suggestion
This area is under rapid development, so the documented is still in progress and subject to change. If you are interested in writing your builder, we encourage you to check [webpack](https://github.com/storybookjs/storybook/tree/next/lib/builder-webpack4), [Vite](https://github.com/storybookjs/builder-vite), and Modern Web's [dev-server-storybook](https://github.com/modernweb-dev/web/blob/master/packages/dev-server-storybook/src/serve/storybookPlugin.ts) source code. In addition, we have a wonderful contributor community on [Storybook Discord](https://discord.gg/storybook) if you have questions. Ping us in the [#contributing](https://discord.com/channels/486522875931656193/839297503446695956) channel.
```

File: docs/builders/webpack.md (hunk 1 above)
Comment (domyen) [reply docs/builders/webpack.md:54]: yea!

File: docs/builders/webpack.md [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+  paths={[
+    'common/storybook-main-webpack5-fsCache.js.mdx',
+  ]}
+/>
+
+<!-- prettier-ignore-end -->
+
+This feature will mean build output is cached between runs of Storybook, speeding up subsequent startup times.
+
+### Extending Storybook’s webpack config
```
Comment (domyen) [inline docs/builders/webpack.md:118]: Please make sure all Webpack's are capitalized when used in a sentence or phrase

**Review by domyen — CHANGES_REQUESTED** [review]

File: docs/builders/builder-api.md (hunk 5 above)
Comment (jonniebigodes) [reply docs/builders/builder-api.md:53]: Actually no,`storybook-static` by default is put on the root level of the project/app. But there could be the case that they can run the build command passing additional CLI options and placing it in another place, they could have set up a monorepo with NX/Turbosap and have multiple Storybook instances each placing them into separate directories. Hence why I left it like this.

**Comment by jonniebigodes** [comment]
Merging this!

**Pull request merged**